### DISCUSSION TITLE
Hoymiles-Library als extra Repository

### DISCUSSION THREAD

**Discussion by sivar2311** [tbnobody/OpenDTU#75] [Ideas]
Moin @tbnobody!
Könntest du dir vorstellen, die Hoymiles-Library in ein extra Repository zu legen?
Damit könnte die Library in andere Projekte einfacher eingebunden werden.
```
lib_deps = 
  https://github.com/tbnobody/hoymiles.git
```

**Comment by battifreund** [comment]
Moin! Das wäre wirklich super. So wäre eine ESPHome-Komponente einfacher möglich.

**Comment by sivar2311** [comment]
Ich habe die Hoymiles Library **zum testen** in ein eigenes Repository unter [sivar2311/Hoymiles](https://github.com/sivar2311/Hoymiles) gelegt und kleine Pfadanpassungen vorgenommen.

Weiterhin habe ich einen [Fork](https://github.com/sivar2311/openDTU) der openDTU mit folgenden Modifikationen angelegt:
- platformio.ini angepasst (`lib_deps` um `https://github.com/sivar2311/Hoymiles.git` ergänzt)
- `ResetReason.h` in das `include` Verzeichnis verschoben
- `ResetReason.cpp`in das `src` Verzeichnis verschoben
- `lib` Verzeichnis gelöscht

Das ganze kompiliert wunderbar und die Library kann / könnte so auch in anderen Projekten wunderbar eingesetzt werden.
Ein kleines Test-Projekt habe ich unter [sivar2311/Hoymiles-Test](https://github.com/sivar2311/HoymilesTest) abgelegt.

Da sich die Pflege dieser externen Library schwierig gestalten wird, werde ich die Repositories demnächst wieder löschen (sie dienen nur zum Test)

**Comment by tbnobody** [comment]
Hallo @sivar2311 ich finde die Idee generell gut die Hoymiles Lib separat zu hosten und einzubinden. Das im aktuellen Stadium zu machen würde aber zum einen die lokale Entwicklung hier etwas verkomplizieren (gut, ich müsste die Lib zum Entwickeln wohl statt über die platformio.ini mit git submodules einbinden) und außerdem ist das Klassen-Interface auch noch alles andere als stabil.

**Comment by sivar2311** [comment]
Hallo @tbnobody !

> Das im aktuellen Stadium zu machen würde aber zum einen die lokale Entwicklung hier etwas verkomplizieren (gut, ich müsste die Lib zum Entwickeln wohl statt über die platformio.ini mit git submodules einbinden)...

Mit git submodules kenne ich mich nicht aus.
Wenn ich Bibliotheken entwickle, erstelle ich zuerst das Repository und clone es dann in das globale PlatformIO Library Verzeichnis unter `C:\Users\<username>\.platformio\lib` (Windows). So kann man prima lokal arbeiten und das (Test-)Projekt greift auf die lokale Version der Library zu. Im Testprojekt dann natürlich ohne den `lib_deps` Eintrag.

> und außerdem ist das Klassen-Interface auch noch alles andere als stabil.

Vielleicht noch nicht stabil, aber dennoch sehr schön umgesetzt!
Da sind noch so einige Tricks die ich mir abschauen kann :)

**Comment by sivar2311** [comment]
Hi @tbnobody !
Ich habe mich mit den git submodules beschäftigt und finde es "suboptimal".
Sowohl beim clonen des Repository als auch beim Download als ZIP-Datei werden die submodules nicht beachtet.
Es würde also die Hoymiles-Bibliothek fehlen und das Projekt lässt sich nicht kompilieren.

Dann doch lieber die Verknüpfung via `lib_deps` :)

**Reply by madmartin** [reply]
@tbnobody 
Ich finde git submodule auch nicht gut. Stimme @sivar2311 in allen Punkten voll zu.

**Comment by tbnobody** [comment]
Hallo @sivar2311 ,

Das herunterladen als ZIP Datei wird sowieso nicht supported bzw. brachte bis vor ein paar commits sowieso einen Fehler beim kompilieren...

Und zum Thema Git Submodules.... https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/How-to-clone-a-git-repository-with-submodules-init-and-update

**Reply by sivar2311** [reply]
`git submodule update` habe ich bei meinen ersten Versuchen auch schon kennen gelernt :)
Leider gibt's m. E. dafür keinen einfachen Menüpunkt in VS-Code.

**Comment by charly4711** [comment]
Moin, ist das Thema damit eigentlich durch?
Bin gerade über das Projekt gestolpert und finde es natürlich toll, dass es eine Alternative zu der Hoymiles Cloud gibt. Andererseits bin ich nach meinen Erfahrungen mit Tasmota jetzt so begeistert von esphome und den Möglichkeiten, das so zu minimieren (WPA3-only, MQTTS, fertig), dass ich ganz ungern jetzt wieder ein anderes Betriebssystem auf ein IoT Device ausbringen würde, wieder mit Web-Seite, API, usw. usw. (zu viele Angriffsvektoren für meinen Geschmack.)
Ich frage mich gerade, was es wohl bräuchte, um z. B. auf einem DX-DTU Board den ESP32 einfach gleich mit esphome zu betanken, und dem halt vorher Support für den NRF24L01+ (Und das Protokoll, das Hoymiles spricht?) beizubringen. Die Library hier scheint mir ein essenzielles Stück Arbeit zu sein, was man nicht nochmal neu erfinden müsste.
